clientupdate/distsign/distsign.go

```diff
@@ -0,0 +1,286 @@
+// distributable files.
+//
+// There are 3 parties in this exchange:
+//   - builder, which creates files, signs them with signing keys and publishes
+//     to server
+//   - server, which distributes public signing keys, files and signatures
+//   - client, which downloads files and signatures from server, and validates
+//     the signatures
+//
+// There are 2 types of keys:
+//   - signing keys, that sign individual distributable files on the builder
+//   - root keys, that sign signing keys and are kept offline
+//
+// root keys -(sign)-> signing keys -(sign)-> files
+//
+// All keys are asymmetric Ed25519 key pairs.
+//
+// The server serves static files under some known prefix. The kinds of files are:
+//   - distsign.pub - bundle of PEM-encoded public signing keys
+//   - distsign.pub.sig - signature of distsign.pub using one of the root keys
+//   - $file - any distributable file
+//   - $file.sig - signature of $file using any of the signing keys
+//
+// The root public keys are baked into the client software at compile time.
+// These keys are long-lived and prove the validity of current signing keys
+// from distsign.pub. To rotate root keys, a new client release must be
+// published, they are not rotated dynamically. There are multiple root keys in
+// different locations specifically to allow this rotation without using the
+// discarded root key for any new signatures.
+//
+// The signing public keys are fetched by the client dynamically before every
+// download and can be rotated more readily, assuming that most deployed
+// clients trust the root keys used to issue fresh signing keys.
+package distsign
+
+import (
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+)
+
+const (
+	pemTypePrivate = "PRIVATE KEY"
+	pemTypePublic  = "PUBLIC KEY"
+
+	downloadSizeLimit    = 1 << 29 // 512Mb
+	signingKeysSizeLimit = 1 << 20 // 1Mb
+	signatureSizeLimit   = ed25519.SignatureSize
+)
+
+// GenerateKey generates a new key pair and encodes it as PEM.
+func GenerateKey() (priv, pub []byte, err error) {
+	pub, priv, err = ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+	return pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePrivate,
+			Bytes: []byte(priv),
+		}), pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePublic,
+			Bytes: []byte(pub),
+		}), nil
+}
+
+// RootKey is a root key Signer used to sign signing keys.
+type RootKey Signer
+
+// SignSigningKeys signs the bundle of public signing keys. The bundle must be
+// a sequence of PEM blocks joined with newlines.
+func (s *RootKey) SignSigningKeys(pubBundle []byte) ([]byte, error) {
+	return s.Sign(nil, pubBundle, crypto.Hash(0))
+}
+
+// SigningKey is a signing key Signer used to sign packages.
+type SigningKey Signer
+
+// SignPackageHash signs a SHA-512 hash of a package.
+func (s SigningKey) SignPackageHash(sha512 []byte) ([]byte, error) {
```

Comment:
I had a tiny thought: Perhaps we should sign the combination of the length of the package in bytes as well as the package's SHA-512. Or, I think equivalently, `SHA512(file size || file contents)` (where `||` is concatenation).

I don't expect that SHA-512 will be meaningfully weakened any time soon, but it would give us just a bit more assurance that we are signing exactly and only the right package. In my limited knowledge, attacks on hash functions (collision or pre-image) typically (but not necessarily always) require finding an input with a different length than the original input has. (For example the MD5 collisions I've seen involve creating colliding inputs by adding additional junk bytes to given files.)

And then we need to specify `binary.LittleEndian` or `BigEndian`. Little-endian seems to have won, so.

Reply:
As for how to decide on signing length || hash vs. hashing length || contents: whatever is easier, I think?

When you use `io.TeeReader` to hash the HTTP response body, you could get the expected length from the Content-Length response header. (We must make sure that's really how many bytes we read and hashed, of course.)

Reply:
Cool, is that a common technique?
We don't want to append the length because most collision attacks append garbage padding?
Appending the length is less error-prone to implement, but I understand if it's less robust.

Reply:
Yeah, I was thinking about appending rather than pre-pending, on the grounds of error-prone-age, and I'm not sure — but appending garbage was indeed the main thing I was afraid of.

Reply:
Added length suffix to signed message. This meant switching from Ed25519ph to regular Ed25519 for package signatures, so we're effectively double (triple?) hashing the input.
Technically for signing keys (where we sign the entire blob instead of a hash) we don't need it, as regular Ed25519 is collision-resistant per [the RFC](https://www.rfc-editor.org/rfc/rfc8032#section-4). But I added the length there too for consistency.

Reply:
If we are worried about SHA-512 we could use blake2 for some diversity. We already use it for tailnet lock so it wont add to the binary size, and ed25519 already hashes the input for sign/verify using sha-512 so we end up with both being used.